cjworkbench/tests/test_i18n_middleware.py

```diff
@@ -0,0 +1,187 @@
+from django.test import RequestFactory, SimpleTestCase
+from django.http import HttpResponse
+from cjworkbench.middleware.i18n import SetCurrentLocaleMiddleware
+from django.contrib.auth.models import AnonymousUser, User
+from cjworkbench.i18n import default_locale
+
+non_default_locale = "el"
+unsupported_locale = "jp"
+
+
+def mock_response(request):
+    response = HttpResponse()
+    response.request = request
+    return response
+
+
+class MockUser:
+    def __init__(self, locale_preference=None):
+        self.is_authenticated = True
+        if locale_preference:
+            self.locale_id = locale_preference
+
+
+class SetCurrentLocaleMiddlewareTest(SimpleTestCase):
+    """
+    Tests the SetCurrentLocaleMiddleware.
+    """
+
+    def setUp(self):
+        self.factory = RequestFactory()
+
+    def _mock_request(
+        self,
+        user=None,
+        session_locale=None,
+        accept_language_header=None,
+        request_locale=None,
+    ):
+        request = self.factory.get(
+            "/" if not request_locale else ("/?locale=" + request_locale)
+        )
+        request.session = {"locale_id": session_locale}
+        request.user = user or AnonymousUser()
+        if accept_language_header:
+            request.META["HTTP_ACCEPT_LANGUAGE"] = accept_language_header
+        return request
+
+    def _process_request(self, **kwargs):
+        return SetCurrentLocaleMiddleware(mock_response)(
+            self._mock_request(**kwargs)
+        ).request
+
+    def test_anonymous_user(self):
+        """ A non-registered user will be served the locale with the following order of preference:
+            1. Current GET request parameter, if the locale is supported
+            2. Session, if the locale is supported
+            3. Accept-Language HTTP header, if the header is valid and contains a supported locale
+            4. Default locale
+        
+        Registered users that have never set a locale preference will be treated as non-registered.
+        """
+        for user in [None, MockUser()]:
+            # 4
+            request = self._process_request(user=user)
```

Comment:
These unit tests are great. Please:

1. Ditch the `MockUser()` case. I'm comfortable with not-testing this behavior because I'm about to add the language field very soon.
2. Write one test function per `request = ...`. e.g., `test_anonymous_user_default_locale`, `test_anonymous_user_invalid_header`, etc. (This is why I'm comfortable dropping the `Mockuser()` case -- so we don't write double the tests.)